File: Backend/src/Fibertest30.Api/Mapping/EventTablesMapping.cs
namespace Fibertest30.Api;

public static class EventTablesMapping
{
    public static HasCurrentEvents ToProto(this Application.HasCurrentEvents hasCurrentEvents)
    {
        return new HasCurrentEvents
        {
            HasCurrentOpticalEvents = hasCurrentEvents.HasCurrentOpticalEvents,
            HasCurrentNetworkEvents = hasCurrentEvents.HasCurrentNetworkEvents,
            HasCurrentBopNetworkEvents = hasCurrentEvents.HasCurrentBopNetworkEvents,
            HasCurrentRtuAccidents = hasCurrentEvents.HasCurrentRtuAccidents,
        };
    }

    private static AccidentNeighbour ToProto(this Iit.Fibertest.Dto.AccidentNeighbourDto dto)
    {
        return new AccidentNeighbour()
        {
            LandmarkIndex = dto.LandmarkIndex,
            Title = dto.Title,
            Coors = dto.Coors.ToProto(),
            ToRtuOpticalDistanceKm = dto.ToRtuOpticalDistanceKm,
            ToRtuPhysicalDistanceKm = dto.ToRtuPhysicalDistanceKm,
        };
    }

    private static AccidentOnTraceV2 ToProto(this Iit.Fibertest.Dto.AccidentOnTraceV2Dto dto)
    {
        AccidentOnTraceV2 accidentOnTraceV2 = new ()
        {
            BrokenRftsEventNumber = dto.BrokenRftsEventNumber,
            AccidentSeriousness = dto.AccidentSeriousness.ToProto(),
            OpticalTypeOfAccident = dto.OpticalTypeOfAccident.ToProto(),

            IsAccidentInOldEvent = dto.IsAccidentInOldEvent,
            IsAccidentInLastNode = dto.IsAccidentInLastNode,
            AccidentCoors = dto.AccidentCoors.ToProto(),

            AccidentLandmarkIndex = dto.AccidentLandmarkIndex,
            AccidentToRtuOpticalDistanceKm = dto.AccidentToRtuOpticalDistanceKm,
            AccidentTitle = dto.AccidentTitle,
            AccidentToRtuPhysicalDistanceKm = dto.AccidentToRtuPhysicalDistanceKm,

            AccidentToLeftOpticalDistanceKm = dto.AccidentToLeftOpticalDistanceKm,
            AccidentToLeftPhysicalDistanceKm = dto.AccidentToLeftPhysicalDistanceKm,
            AccidentToRightOpticalDistanceKm = dto.AccidentToRightOpticalDistanceKm,
            AccidentToRightPhysicalDistanceKm = dto.AccidentToRightPhysicalDistanceKm,

            EventCode = dto.EventCode,
            DeltaLen = dto.DeltaLen,
        };

        if (dto.Left != null)
            accidentOnTraceV2.Left = dto.Left.ToProto();
        if (dto.Right != null)
            accidentOnTraceV2.Right = dto.Right.ToProto();

        return accidentOnTraceV2;
    }

    public static OpticalEvent ToProto(this Iit.Fibertest.Dto.OpticalEventDto dto)
    {
        // файбертест исторически хранит все даты в локальном времени
        // перед передачей по grpc переводим сначала время в UTC, а затем в прото формат
        // вэб морда получает все времена UTC и
        // DateTimePipe взяв из браузера таймзону клиента показывает время правильно для каждого клиента


        OpticalEvent opticalEvent = new OpticalEvent
        {
            EventId = dto.EventId,
            MeasuredAt = dto.MeasurementTimestamp.ToUniversalTime().ToTimestamp(),
            RegisteredAt = dto.EventRegistrationTimestamp.ToUniversalTime().ToTimestamp(),

            RtuTitle = dto.RtuTitle,
            RtuId = dto.RtuId.ToString(),
            TraceTitle = dto.TraceTitle,
            TraceId = dto.TraceId.ToString(),

            BaseRefType = dto.BaseRefType.ToProto(),
            TraceState = dto.TraceState.ToProto(),

            EventStatus = dto.EventStatus.ToProto(),
            StatusChangedAt = dto.StatusChangedTimestamp.ToUniversalTime().ToTimestamp(),
            StatusChangedByUser = dto.StatusChangedByUser,

            Comment = dto.Comment
        };

        dto.Accidents.ForEach(a=> opticalEvent.Accidents.Add(a.ToProto()));
        return opticalEvent;
    }

    public static NetworkEvent ToProto(this Iit.Fibertest.Dto.NetworkEventDto dto)
    {
        return new NetworkEvent
        {
            EventId = dto.EventId,
            RegisteredAt = dto.EventRegistrationTimestamp.ToUniversalTime().ToTimestamp(),
            RtuId = dto.RtuId.ToString(),
            RtuTitle = dto.RtuTitle,
            IsRtuAvailable = dto.IsRtuAvailable,
            OnMainChannel = dto.OnMainChannel.ToProto(),
            OnReserveChannel = dto.OnReserveChannel.ToProto(),
        };
    }

    public static BopEvent ToProto(this Iit.Fibertest.Dto.BopEventDto dto)
    {
        return new BopEvent
        {
            EventId = dto.EventId, 
            RegisteredAt = dto.EventRegistrationTimestamp.ToUniversalTime().ToTimestamp(), 
            BopAddress = dto.BopAddress,
            RtuId = dto.RtuId.ToString(),
            RtuTitle = dto.RtuTitle,
            Serial = dto.Serial,
            IsBopOk = dto.BopState,
        };
    }

    public static RtuAccident ToProto(this Iit.Fibertest.Dto.RtuAccidentDto dto)
    {
        return new RtuAccident
        {
            Id = dto.Id,
            IsMeasurementProblem = dto.IsMeasurementProblem,
            ReturnCode = (int)dto.ReturnCode,
            RegisteredAt = dto.EventRegistrationTimestamp.ToUniversalTime().ToTimestamp(),
            RtuTitle = dto.RtuTitle,
            RtuId = dto.RtuId.ToString(),
            TraceTitle = dto.TraceTitle,
            TraceId = dto.TraceId.ToString(),
            BaseRefType = dto.BaseRefType.ToProto(),
            Comment = dto.Comment,
            ClearedAccidentWithId = dto.ClearedAccidentWithId
        };
    }
}

File: WebFt30/src/app/core/store/mapping/event-tables-mapping.ts
import * as grpc from 'src/grpc-generated';
import { HasCurrentEvents } from '../models/ft30/has-current-events';
import { BopEvent } from '../models/ft30/bop-event';
import { Timestamp } from 'src/grpc-generated/google/protobuf/timestamp';
import { AccidentNeighbour, AccidentOnTraceV2, OpticalEvent } from '../models/ft30/optical-event';
import { FtEnumsMapping } from './ft-enums-mapping';
import { NetworkEvent } from '../models/ft30/network-event';
import { RtuAccident } from '../models/ft30/rtu-accident';
import { GisMapping } from './gis-mappings';

export class EventTablesMapping {
  static fromGrpcHasCurrentEvents(grpcHas: grpc.HasCurrentEvents): HasCurrentEvents {
    const result = new HasCurrentEvents();
    result.optical = grpcHas.hasCurrentOpticalEvents;
    result.networkRtu = grpcHas.hasCurrentNetworkEvents;
    result.networkBop = grpcHas.hasCurrentBopNetworkEvents;
    result.rtuAccidents = grpcHas.hasCurrentRtuAccidents;
    return result;
  }

  static toOpticalEvents(grpcOpticalEvents: grpc.OpticalEvent[]): OpticalEvent[] {
    const opticalEvents = grpcOpticalEvents.map((item) => this.toOpticalEvent(item));
    return opticalEvents;
  }

  static toOpticalEvent(grpcOpticalEvent: grpc.OpticalEvent): OpticalEvent {
    const opticalEvent = new OpticalEvent();
    opticalEvent.eventId = grpcOpticalEvent.eventId;
    opticalEvent.measuredAt = Timestamp.toDate(grpcOpticalEvent.measuredAt!);
    opticalEvent.registeredAt = Timestamp.toDate(grpcOpticalEvent.registeredAt!);

    opticalEvent.rtuTitle = grpcOpticalEvent.rtuTitle;
    opticalEvent.rtuId = grpcOpticalEvent.rtuId;
    opticalEvent.traceTitle = grpcOpticalEvent.traceTitle;
    opticalEvent.traceId = grpcOpticalEvent.traceId;

    opticalEvent.baseRefType = grpcOpticalEvent.baseRefType;
    opticalEvent.traceState = FtEnumsMapping.fromGrpcFiberState(grpcOpticalEvent.traceState);

    opticalEvent.eventStatus = grpcOpticalEvent.eventStatus;
    opticalEvent.statusChangedAt = Timestamp.toDate(grpcOpticalEvent.statusChangedAt!);
    opticalEvent.statusChangedByUser = grpcOpticalEvent.statusChangedByUser;

    opticalEvent.comment = grpcOpticalEvent.comment;
    opticalEvent.accidents = grpcOpticalEvent.accidents.map((a) => this.toAccidentOnTraceV2(a));
    return opticalEvent;
  }

  static toAccidentOnTraceV2(grpcAccident: grpc.AccidentOnTraceV2): AccidentOnTraceV2 {
    const accidentOnTrace = new AccidentOnTraceV2();
    accidentOnTrace.brokenRftsEventNumber = grpcAccident.brokenRftsEventNumber;

    accidentOnTrace.accidentSeriousness = FtEnumsMapping.fromGrpcFiberState(
      grpcAccident.accidentSeriousness
    );
    accidentOnTrace.opticalTypeOfAccident = FtEnumsMapping.fromGrpcOpticalAccidentType(
      grpcAccident.opticalTypeOfAccident
    );

    accidentOnTrace.isAccidentInOldEvent = grpcAccident.isAccidentInOldEvent;
    accidentOnTrace.isAccidentInLastNode = grpcAccident.isAccidentInLastNode;
    accidentOnTrace.accidentCoors = GisMapping.fromGeoCoordinate(grpcAccident.accidentCoors!);

    accidentOnTrace.accidentLandmarkIndex = grpcAccident.accidentLandmarkIndex;
    accidentOnTrace.accidentToRtuOpticalDistanceKm = grpcAccident.accidentToRtuOpticalDistanceKm;
    accidentOnTrace.accidentTitle = grpcAccident.accidentTitle;
    accidentOnTrace.accidentToRtuPhysicalDistanceKm = grpcAccident.accidentToRtuPhysicalDistanceKm;

    accidentOnTrace.accidentToLeftOpticalDistanceKm = grpcAccident.accidentToLeftOpticalDistanceKm;
    accidentOnTrace.accidentToLeftPhysicalDistanceKm =
      grpcAccident.accidentToLeftPhysicalDistanceKm;
    accidentOnTrace.accidentToRightOpticalDistanceKm =
      grpcAccident.accidentToRightOpticalDistanceKm;
    accidentOnTrace.accidentToRightPhysicalDistanceKm =
      grpcAccident.accidentToRightPhysicalDistanceKm;

    accidentOnTrace.eventCode = grpcAccident.eventCode;
    accidentOnTrace.deltaLen = grpcAccident.deltaLen;

    if (grpcAccident.left) accidentOnTrace.left = this.toAccidentNeighbour(grpcAccident.left);
    if (grpcAccident.right) accidentOnTrace.right = this.toAccidentNeighbour(grpcAccident.right);
    return accidentOnTrace;
  }

  static toAccidentNeighbour(grpcNeighbour: grpc.AccidentNeighbour): AccidentNeighbour {
    const accidentNeighbour = new AccidentNeighbour();
    accidentNeighbour.landmarkIndex = grpcNeighbour.landmarkIndex;
    accidentNeighbour.title = grpcNeighbour.title;
    accidentNeighbour.coors = GisMapping.fromGeoCoordinate(grpcNeighbour.coors!);
    accidentNeighbour.toRtuOpticalDistanceKm = grpcNeighbour.toRtuOpticalDistanceKm;
    accidentNeighbour.toRtuPhysicalDistanceKm = grpcNeighbour.toRtuPhysicalDistanceKm;
    return accidentNeighbour;
  }

  static toNetworkEvents(grpcNetworkEvents: grpc.NetworkEvent[]): NetworkEvent[] {
    const networkEvents = grpcNetworkEvents.map((item) => this.toNetworkEvent(item));
    return networkEvents;
  }

  static toNetworkEvent(grpcNetworkEvent: grpc.NetworkEvent): NetworkEvent {
    const networkEvent = new NetworkEvent();
    networkEvent.eventId = grpcNetworkEvent.eventId;
    networkEvent.registeredAt = Timestamp.toDate(grpcNetworkEvent.registeredAt!);
    networkEvent.rtuTitle = grpcNetworkEvent.rtuTitle;
    networkEvent.rtuId = grpcNetworkEvent.rtuId;
    networkEvent.isRtuAvailable = grpcNetworkEvent.isRtuAvailable;
    networkEvent.onMainChannel = FtEnumsMapping.fromGrpcChannelEvent(
      grpcNetworkEvent.onMainChannel
    );
    networkEvent.onReserveChannel = FtEnumsMapping.fromGrpcChannelEvent(
      grpcNetworkEvent.onReserveChannel
    );
    return networkEvent;
  }

  static toBopEvents(grpcBopEvents: grpc.BopEvent[]): BopEvent[] {
    return grpcBopEvents.map((i) => this.toBopEvent(i));
  }

  static toBopEvent(grpcBopEvent: grpc.BopEvent): BopEvent {
    const bopEvent = new BopEvent();
    bopEvent.eventId = grpcBopEvent.eventId;
    bopEvent.registeredAt = Timestamp.toDate(grpcBopEvent.registeredAt!);
    bopEvent.bopAddress = grpcBopEvent.bopAddress;
    bopEvent.rtuTitle = grpcBopEvent.rtuTitle;
    bopEvent.rtuId = grpcBopEvent.rtuId;
    bopEvent.serial = grpcBopEvent.serial;
    bopEvent.isBopOk = grpcBopEvent.isBopOk;
    return bopEvent;
  }

  static toRtuAccidents(grpcRtuAccidents: grpc.RtuAccident[]): RtuAccident[] {
    return grpcRtuAccidents.map((a) => this.toRtuAccident(a));
  }

  static toRtuAccident(grpcRtuAccident: grpc.RtuAccident): RtuAccident {
    const rtuAccident = new RtuAccident();
    rtuAccident.id = grpcRtuAccident.id;
    rtuAccident.isMeasurementProblem = grpcRtuAccident.isMeasurementProblem;
    rtuAccident.returnCode = FtEnumsMapping.fromGrpcReturnCode(grpcRtuAccident.returnCode);
    rtuAccident.registeredAt = Timestamp.toDate(grpcRtuAccident.registeredAt!);
    rtuAccident.rtuTitle = grpcRtuAccident.rtuTitle;
    rtuAccident.rtuId = grpcRtuAccident.rtuId;
    rtuAccident.traceTitle = grpcRtuAccident.traceTitle;
    rtuAccident.traceId = grpcRtuAccident.traceId;
    rtuAccident.baseRefType = grpcRtuAccident.baseRefType;
    rtuAccident.comment = grpcRtuAccident.comment;
    rtuAccident.clearedAccidentWithId = grpcRtuAccident.clearedAccidentWithId;

    return rtuAccident;
  }
}
